### ISSUE TITLE
proposal: `null_aware_followed_by_null_assert`

### ISSUE THREAD

**Issue by srawlins** [dart-lang/linter#4976]
## Description

This one is a little odd, but I've seen a few cases. In code like this:

```dart
    foobarDateRange = (foobarDateRange
          .clampSomething(Something.VALUE)!
          .applySomething()
          ?.clampToSomething(Something.OTHER_VALUE)!)!;
```

It's a bit hard to follow the parens and bangs, but that last line essentially has `?.foo(bar)!` in it. I don't think there is ever any reason to have a null-aware method call or null aware property access (or chain of these) followed by a null-assert. It fails in the same way as `!.foo(bar)`, with a more-or-equally informative stack trace, and is more readable.

## Kind

Readability

## Bad Examples

```dart
a.b?.c!;
a.b?.c()!;
a.b?.c.d().e!;
```

Not sure if this needs to be careful around cascades...

## Good Examples 

*Add a few examples that demonstrate a “good” adoption of this lint’s principle.*

[Writing Lints]: https://github.com/dart-lang/linter/blob/main/doc/writing-lints.md
[Effective Dart]: https://dart.dev/effective-dart
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[SDK Tracker]: https://github.com/dart-lang/sdk/issues

**Comment by bwilkerson** [comment]
What about cases where the chain after the `?` can also introduce a null? For example, if `foo` can return `null`, is `?.foo(bar)!` better than `!.foo(bar)!`?

**Comment by srawlins** [comment]
Right. It is not. A null-assert that throws an exception in `?.foo(bar)!` is either thrown because `foo()` returned `null` or because null-shortening from the `?.` got us to the null-assert.

If the second null-assert throws an exception in `!.foo(bar)!`, it is because `foo()` returned `null`. If the first throws an exception, it is because the value to the left of the code in question returned `null`.
